csrc/quantization/quant_llm/quant_llm_linear.cu

```diff
@@ -0,0 +1,302 @@
+  size_t dimM = M_Global * Split_K / TilingConfig::TILE_M;
+  dim3 GridDim(dimN, dimM, 1);
+  dim3 BlockDim(WARP_SIZE * TilingConfig::BLOCK_WARPS, 1, 1);
+//
+#ifdef DEBUG_MODE
+  printf(
+      "GridDim.x: %d, GridDim.y: %d, GridDim.z: %d, BlockDim.x: %d, "
+      "BlockDim.y: %d, BlockDim.z: %d SHMEM_SZ: %d\n",
+      GridDim.x, GridDim.y, GridDim.z, BlockDim.x, BlockDim.y, BlockDim.z,
+      SHMEM_SZ);
+  printf("\n");
+#endif
+  QUANT_GEMM_Kernel<TilingConfig, OutputDataType, EXPONENT, MANTISSA>
+      <<<GridDim, BlockDim, SHMEM_SZ, stream>>>(Weight, Scales, B, C, M_Global,
+                                                N_Global, K_Global, Split_K);
+}
+
+template <int EXPONENT, int MANTISSA>
+cudaError_t fpx_linear_kernel(
+    cudaStream_t stream, const uint4* Weight, const half* Scales, const half* B,
+    half* C, const size_t M_Global, const size_t N_Global,
+    const size_t K_Global,
+    float* Reduction_Workspace,  // Reduction_Workspace_Size = Split_K *
+                                 // M_Global * N_Global * sizeof(fp32)
+    int Split_K) {
+  assert(M_Global % 256 == 0);
+  assert(K_Global % 64 == 0);
+  assert(N_Global > 0);
+
+  // Work around to support more N shapes:
+  size_t N_PowerOf2;
+  if (N_Global > 0 && N_Global <= 8) N_PowerOf2 = 8;
+  if (N_Global > 8 && N_Global <= 16) N_PowerOf2 = 16;
+  if (N_Global > 16 && N_Global <= 32) N_PowerOf2 = 32;
+  if (N_Global > 32 && N_Global <= 64) N_PowerOf2 = 64;
+  if (N_Global > 64 && N_Global <= 128) N_PowerOf2 = 128;
+  if (N_Global > 128) N_PowerOf2 = ((N_Global - 1) / 128 + 1) * 128;
+
+  if (Split_K == 1) {
+    switch (N_PowerOf2) {
+      case 8:
+        Kernel_Ex<TilingConfig<4, 1, 1>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
+        break;
+      case 16:
+        Kernel_Ex<TilingConfig<4, 1, 2>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
+        break;
+      case 32:
+        Kernel_Ex<TilingConfig<4, 1, 4>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
+        break;
+      case 64:
+        Kernel_Ex<TilingConfig<4, 1, 8>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
+        break;
+      case 128:
+        Kernel_Ex<TilingConfig<4, 1, 8>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
+        break;
+      default:
+        if (N_PowerOf2 % 128 != 0) {
+          printf("FP6LLM_API Error: Unsupported N dimension %d!\n", N_PowerOf2);
+          return cudaErrorUnknown;
+        }
+        Kernel_Ex<TilingConfig<4, 1, 8>, half, EXPONENT, MANTISSA>(
+            stream, Weight, Scales, B, C, M_Global, N_Global, K_Global,
+            Split_K);
```

Comment:
Why is there a function here if we are early exiting? Also we usually use TORCH_CHECK or TORCH_CHECK_NOT_IMPLEMENTED for asserts

Reply:
The default case for the execution flow is `N_PowerOf2` being a multiple of 128, so this should be fine here. Do you have any concerns about this?

I'll switch to using TORCH_CHECK for the assertions.